File: Construct Binary Tree from Inorder and Postorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        // Start typing your Java solution below
        // DO NOT write main() function
        HashMap<Integer, Integer> indexMap = new HashMap<Integer, Integer>();
        for(int i = 0; i < inorder.length; i ++)
        {
            indexMap.put(inorder[i], i);
        }
        
        return buildTree(postorder, postorder.length, 0, 0, indexMap);
    }
    
    public TreeNode buildTree(int []postorder, int size, int start, int offset, HashMap<Integer, Integer> indexMap)
    {
        if(size < 1)
        {
            return null;
        }
        
        int lastIndex = start + size - 1;
        int val = postorder[lastIndex];
        int leftSize = indexMap.get(val) - offset;
        int rightSize = size - leftSize - 1;
        TreeNode root = new TreeNode(val);
        root.left = buildTree(postorder, leftSize, start, offset, indexMap);
        root.right = buildTree(postorder, rightSize, start + leftSize, offset + leftSize + 1, indexMap);
        return root;
    }
}

File: Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  unordered_map<int, int> indexMap;
    
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        //this->preorder = preorder;
        //this->inorder = inorder;
        
        int N = inorder.size();
        indexMap.clear();
        for(int i = 0; i < N; i++)
        {
            indexMap[inorder[i]] = i;
        }
        
        return buildTree(postorder, 0, N, 0);
    }
    
    TreeNode *buildTree(vector<int> &postorder, int postorderStart, int N, int offset)
    {
        if(N<=0)
            return NULL;
        
        int val = postorder[N-1+postorderStart];
        int inorderIndex = indexMap[val];
        TreeNode *root = new TreeNode(val);
        int i = inorderIndex - offset;
        
        root->left = buildTree(postorder,postorderStart, i, offset);
        root->right = buildTree(postorder, postorderStart+i, N-i-1, offset + i + 1);
        
        return root;
    }
};